### DISCUSSION TITLE
`atomWithStorage` Add option to modify serialize and deserialize functions

### DISCUSSION THREAD

**Discussion by ViktorQvarfordt** [pmndrs/jotai#2321] [Ideas] [answered]
The example for [validating stored values](https://jotai.org/docs/utilities/storage#validating-stored-values) is quite complex. One shouldn't have to reimplement the entire storage just to modify serialize and deserialize, which I believe to be the most common use case. I suggest something like this:

```ts
const myAtom = atomWithStorage('my-key', null, {
  serialize: (value) => JSON.stringify(value),
  deserialize: (str) => JSON.parse(str)
})
```

I want to use all the other defaults of the localStorage implementation. I just need to modify the serialize and deserialize (because I want to persist an atom of atoms).

**Comment by dai-shi** [comment]
Our design decision is to provide utils to help creating a storage, instead of extending atomWithStorage.

There are several options:
1. One idea is to allow `reviver` and `replacer` options in createJSONStorage.
2. If it's beyond simple JSON.parse/stringify, we should create a new one, hopefully with validation (which is recommended in production. In that sense, createJSONStorage isn't recommended.)
3. Some validation mechanisms might depend on 3rd-party libs. In such a case, we want to provide utils by a separate library.

These options are not exclusive. We can do them all.

**Comment by dai-shi** [comment]
> to modify serialize and deserialize, which I believe to be the most common use case.

What is your use case?

**Comment by ViktorQvarfordt** [comment]
Use cases:
1. Validating data after deserialization.
2. Persisting an atom of atoms (custom serialization and deserialization).
3. Serializing and deserializing non-json data such as Maps and Sets (custom serialization and deserialization).

**Reply by dai-shi** [reply]
1: What should be done if data is not validated?
2,3: Should be done with `reviver` and `replacer`.

**Reply by ViktorQvarfordt** [reply]
1. That's up to the implementation of the deserialization function. I would drop invalid data and, if the data is high stakes, notify the user and show them the invalid data.
2. I'd say the names serialize and deserialize are more common and self-explanatory than reviver and replacer. I don't know how to understand reviver; what is being revived, and from where?

**Reply by dai-shi** [reply]
Sorry, it's out of context. I mean to modify createJSONStorage with the following options.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#reviver
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#replacer

**Reply by dai-shi** [reply]
#2324

**Reply by ViktorQvarfordt** [reply]
I see, thanks. But assuming that JSON.parse/stringify is to be used is perhaps a too strong assumption. I might want to serialize to text in a different way.

**Reply by dai-shi** [reply]
Yes, I can assume one might want to store as just strings.
We should use storage option for it without createJSONStorage, which is for JSON.parse/stringify. (maybe we can consider `createCustomStroage` that doesn't assume JSON.)

At least, revivier/replacer should solve your use case 3.

**Comment by ViktorQvarfordt** [comment]
I'd rather see an easy way to modify the serialize and deserialize methods than integrations/plugins for each particular use case.

**Reply by dai-shi** [reply]
I think the idea is simply conflicting.
But, looking at the docs again, I see it's complicated, mostly because it shows all possible methods. There seems room for improvement in docs to show more easier use cases.

**Reply by ViktorQvarfordt** [reply]
Is there already a way to just modify the serialization and deserialization steps?

I understand now that these might not be options to atomWithStorage itself, but perhaps options to an existing (configurable) storage implementation compatible with atomWithStorage.

**Comment by ViktorQvarfordt** [comment, marked as answer]
I see you updated the lib and the docs. Thanks!

I've seen my needs be best handled by this custom implementation that I'm using instead now:

```ts
type SetStateActionWithReset<Value> = Value | typeof RESET | ((prev: Value) => Value | typeof RESET)

export const atomWithCustomStorage = <T>({
  key,
  initialValue,
  serialize,
  deserialize,
}: {
  key: string
  initialValue: T
  serialize: (val: T) => string
  deserialize: (str: string) => T
}): PrimitiveAtom<T> => {
  const str = localStorage.getItem(key)

  let val = initialValue

  if (str !== null) {
    try {
      val = deserialize(str)
    } catch (err) {
      console.warn(`Failed to deserialize localStorage.getItem('${key}'). Using initial value.`)
    }
  }

  const baseAtom = atom(val)

  return atom(
    get => get(baseAtom),
    (get, set, update: SetStateActionWithReset<T>) => {
      // @ts-expect-error "This expression is not callable." https://github.com/microsoft/TypeScript/issues/37663
      const newVal = typeof update === 'function' ? update(get(baseAtom)) : update

      if (newVal === RESET) {
        set(baseAtom, initialValue)
        localStorage.removeItem(key)
      } else {
        set(baseAtom, newVal)
        localStorage.setItem(key, serialize(newVal))
      }
    }
  )
}
```
